activerecord/lib/active_record/migration/compatibility.rb

```diff
@@ -54,6 +54,10 @@ def add_column(table_name, column_name, type, **options)
           super
         end
 
+        def add_index(table_name, column_name, **options)
```

Comment:
I haven't re-read enough context to check, but can this be `def generate_index_name(table_name, column)` with the body from `legacy_index_name`?

"Hidden" options like this are an available fallback when all other options fail, but they should be a last resort: they push the legacy behaviour into the modern implementation instead of isolating it in the compatibility handling -- plus it introduces a singular "legacy" state, losing the version-specific scoping we start with.

Reply:
That would be much nicer! I originally tried that approach. It might be possible but I was struggling to find a way to make it work.

The challenge was, it looks to me like the `migrator` will call the overridden method if available (such as `add_index`). If not, it falls back to method_missing.

When I tried overriding `generate_index_name` from Compatibility, it was never being called. I believe because the migrator is never directly calling it.

**Possible idea:**
Maybe I could do something like...

```ruby
# in compatibility
def add_index(...)
  options[:name] = generate_index_name(...) if options[:name].nil?
  super
end
```

Then it would be setting the name explicitly using the "legacy" method as if that is what the user passed in.

I can play around with this more later, just sharing the context and wondering your opinion. 😄 Thanks!

Reply:
Got it working! No more hidden option.